File: backend/routes/dosen.py
from flask import Blueprint
from flask_jwt_extended import get_jwt_identity

from models import Dosen, Mahasiswa
from utils.decorators import dosen_required
from utils.helpers import success_response, error_response
from utils.cache import get_cache, set_cache

dosen_bp = Blueprint("dosen", __name__)


@dosen_bp.route("/profile", methods=["GET"])
@dosen_required
def get_profile():
    dosen_id = get_jwt_identity()
    dosen = Dosen.query.filter_by(id=dosen_id).first()
    if not dosen:
        return error_response("Dosen tidak ditemukan", 404)
    return success_response(dosen.to_dict())


@dosen_bp.route("/mahasiswa-bimbingan", methods=["GET"])
@dosen_required
def get_mahasiswa_bimbingan():
    dosen_id  = get_jwt_identity()
    cache_key = f"krs:dosen:{dosen_id}:bimbingan"

    cached = get_cache(cache_key)
    if cached:
        return success_response(cached)

    mahasiswa_list = Mahasiswa.query.filter_by(dosen_pa_id=dosen_id).all()

    result = []
    for mhs in mahasiswa_list:
        mk_list = mhs.mata_kuliah or []
        entry   = mhs.to_dict()
        entry["ringkasan_krs"] = {
            "jumlah_mk": len(mk_list),
            "total_sks": sum(mk.get("sks", 0) for mk in mk_list),
        }
        result.append(entry)

    set_cache(cache_key, result, ttl=300)
    return success_response(result)


@dosen_bp.route("/mahasiswa-bimbingan/<uuid:mahasiswa_id>", methods=["GET"])
@dosen_required
def get_detail_mahasiswa(mahasiswa_id):
    dosen_id = get_jwt_identity()

    mhs = Mahasiswa.query.filter_by(id=mahasiswa_id).first()
    if not mhs:
        return error_response("Mahasiswa tidak ditemukan", 404)

    if str(mhs.dosen_pa_id) != dosen_id:
        return error_response("Mahasiswa ini bukan bimbingan Anda", 403)

    mk_list = mhs.mata_kuliah or []
    data    = mhs.to_dict()
    data["krs"] = {
        "mata_kuliah": mk_list,
        "total_sks":   sum(mk.get("sks", 0) for mk in mk_list),
        "jumlah_mk":   len(mk_list),
    }

    return success_response(data)

File: backend-golang/handlers/dosen.go
package handlers

import (
	"net/http"

	"github.com/google/uuid"
	"github.com/labstack/echo/v4"
	"gorm.io/gorm"

	"krs/backend/models"
	"krs/backend/services"
	"krs/backend/utils"
)

type DosenHandler struct {
	DB    *gorm.DB
	Cache *utils.CacheClient
}

func NewDosenHandler(db *gorm.DB, cache *utils.CacheClient) *DosenHandler {
	return &DosenHandler{DB: db, Cache: cache}
}

func (h *DosenHandler) Profile(c echo.Context) error {
	dosenID, err := uuid.Parse(c.Get("userID").(string))
	if err != nil {
		return utils.ErrorResponse(c, "User ID tidak valid", http.StatusBadRequest)
	}

	var dosen models.Dosen
	if err := h.DB.Where("id = ?", dosenID).First(&dosen).Error; err != nil {
		return utils.ErrorResponse(c, "Dosen tidak ditemukan", http.StatusNotFound)
	}
	return utils.SuccessResponse(c, dosen.ToResponse(), "Success", http.StatusOK)
}

func (h *DosenHandler) GetMahasiswaBimbingan(c echo.Context) error {
	dosenID, err := uuid.Parse(c.Get("userID").(string))
	if err != nil {
		return utils.ErrorResponse(c, "User ID tidak valid", http.StatusBadRequest)
	}

	cacheKey := "krs:dosen:" + dosenID.String() + ":bimbingan"
	if cached := h.Cache.GetCache(cacheKey); cached != nil {
		return utils.SuccessResponse(c, cached, "Success", http.StatusOK)
	}

	result, err := services.GetMahasiswaBimbingan(h.DB, dosenID)
	if err != nil {
		return utils.ErrorResponse(c, "Gagal mengambil data mahasiswa bimbingan", http.StatusInternalServerError)
	}

	h.Cache.SetCache(cacheKey, result, 300)
	return utils.SuccessResponse(c, result, "Success", http.StatusOK)
}

func (h *DosenHandler) GetDetailMahasiswa(c echo.Context) error {
	dosenID, err := uuid.Parse(c.Get("userID").(string))
	if err != nil {
		return utils.ErrorResponse(c, "User ID tidak valid", http.StatusBadRequest)
	}

	mahasiswaID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		return utils.ErrorResponse(c, "ID tidak valid", http.StatusBadRequest)
	}

	data, err := services.GetDetailMahasiswaBimbingan(h.DB, dosenID, mahasiswaID)
	if err != nil {
		switch err.Error() {
		case "not found":
			return utils.ErrorResponse(c, "Mahasiswa tidak ditemukan", http.StatusNotFound)
		case "forbidden":
			return utils.ErrorResponse(c, "Mahasiswa ini bukan bimbingan Anda", http.StatusForbidden)
		}
		return utils.ErrorResponse(c, "Gagal mengambil detail mahasiswa", http.StatusInternalServerError)
	}
	return utils.SuccessResponse(c, data, "Success", http.StatusOK)
}
